syntax = "proto3";

package cakelab.arflow_grpc.v1;

import "cakelab/arflow_grpc/v1/vector2_int.proto";

option csharp_namespace = "CakeLab.ARFlow.Grpc.V1";

/// https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@6.0/api/UnityEngine.XR.ARSubsystems.XRCpuImage.html
message XRCpuImage {
  /// https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@6.0/api/UnityEngine.XR.ARSubsystems.XRCpuImage.Format.html
  enum Format {
    FORMAT_UNSPECIFIED = 0;
    // @exclude The number in each field should match the enum XRCpuImage.Format for more convenient conversion.
    FORMAT_ANDROID_YUV_420_888 = 1;
    FORMAT_IOS_YP_CBCR_420_8BI_PLANAR_FULL_RANGE = 2;
    FORMAT_ONECOMPONENT8 = 3;
    FORMAT_DEPTHFLOAT32 = 4; /// iOS
    FORMAT_DEPTHUINT16 = 5; /// Android
    FORMAT_ONECOMPONENT32 = 6;
    FORMAT_ARGB32 = 7;
    FORMAT_RGBA32 = 8;
    FORMAT_BGRA32 = 9;
    FORMAT_RGB24 = 10;
    FORMAT_H264RGB24 = 16;
  }

  /// https://docs.unity3d.com/Packages/com.unity.xr.arfoundation@6.0/api/UnityEngine.XR.ARSubsystems.XRCpuImage.Plane.html
  message Plane {
    int32 row_stride = 1;
    int32 pixel_stride = 2;
    bytes data = 3;
  }

  Vector2Int dimensions = 1;
  Format format = 2;
  double timestamp = 3;
  repeated Plane planes = 4;
}
